gridappsd/simulation.py

```diff
@@ -1,63 +1,140 @@
+
+        if 'simulationId' not in resp:
+            message = "Simulation was not able to run\n" + str(resp)
+            raise Exception(message)
+        self.simulation_id = resp['simulationId']
+        # Subscribe to the different components necessary to run and receive
+        # simulated measurements and messages.
+        self._gapps.subscribe(t.simulation_log_topic(self.simulation_id), self.__on_simulation_log)
+        self._gapps.subscribe(t.simulation_output_topic(self.simulation_id), self.__onmeasurement)
+        self._gapps.subscribe(t.platform_log_topic(), self.__on_platformlog)
+
+        for p in self.__on_start:
+            p(self)
 
     def pause(self):
-        """Pause simulation"""
+        """ Pause simulation"""
+        _log.debug("Pausing simulation")
         command = dict(command="pause")
-        self.gappsd.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+        self._gapps.send(simulation_input_topic(self.simulation_id), json.dumps(command))
 
     def stop(self):
-        """Stop the simulation"""
+        """ Stop the simulation"""
+        _log.debug("Stopping simulation")
         command = dict(command="stop")
-        self.gappsd.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+        self._gapps.send(simulation_input_topic(self.simulation_id), json.dumps(command))
 
     def resume(self):
-        """Resume the simulation"""
+        """ Resume the simulation"""
+        _log.debug("Resuming simulation")
         command = dict(command="resume")
-        self.gappsd.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+        self._gapps.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+
+    def add_onmesurement_callback(self, callback, device_filter=()):
+        self.__filterable_measurement_callback_set.add((callback, device_filter))
+
+    def add_onstart_callback(self, callback):
+        self.__on_start.add(callback)
+
+    def add_oncomplete_callback(self, callback):
+        self.__on_simulation_complete_callbacks.add(callback)
+
+    def add_ontimestep_callback(self, callback):
+        self.__on_next_timestep_callbacks.add(callback)
+
+    def __on_platformlog(self, headers, message):
+        if self.simulation_id == message['processId']:
+            _log.debug("__on_platformlog: {}".format(message))
+
+        if 'command' in message:
+            _log.debug("Command was: {}".format(message))
+
+    def __on_simulation_log(self, headers, message):
+        # Handle the callbacks here
+        if 'logMessage' in message:
+            log_message = message['logMessage']
+            # if this is the last timestamp then call the finished callbacks
+            if log_message == "incrementing to {}".format(int(self._num_timesteps)):
+                for p in self.__on_simulation_complete_callbacks:
+                    p(self)
```

Comment:
@craig8 Why are we passing `self` to `p`? Shouldn't we be passing through the headers and message? I'm not exactly sure what the intent is here.